elasticdl/pkg/ps/optimizer_test.go

```diff
@@ -63,22 +63,104 @@ func TestSGDOptimizer(t *testing.T) {
 
 	vectors := p.GetEmbeddingParam("t3").GetEmbeddingVectors(i3)
 	assert.Equal(t, 2, len(vectors))
-	assert.True(t, common.CompareFloat(-0.1, vectors[0][0], 0.0001))
-	assert.True(t, common.CompareFloat(-0.1, vectors[1][0], 0.0001))
+	assert.True(t, common.CompareFloat(-0.1, vectors[0].Value[0], 0.0001))
+	assert.True(t, common.CompareFloat(-0.1, vectors[1].Value[0], 0.0001))
 
 	// more test for sparse parameter update
 	d3 = []int64{4, 2}
 	v3 = []float32{1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0}
 	i3 = []int64{1, 3, 3, 5}
 	grad3 = common.Tensor{"t3", v3, d3, i3}
-	grads = []common.Tensor{grad3}
+	grads = []*common.Tensor{&grad3}
 
 	err4 := opt.ApplyGradients(grads, p)
 	assert.Nil(t, err4)
 
 	vectors = p.GetEmbeddingParam("t3").GetEmbeddingVectors(i3)
 	assert.Equal(t, 4, len(vectors))
-	assert.True(t, common.CompareFloat(-0.2, vectors[0][0], 0.0001))
-	assert.True(t, common.CompareFloat(-0.3, vectors[1][0], 0.0001))
-	assert.True(t, common.CompareFloat(-0.1, vectors[3][0], 0.0001))
+	assert.True(t, common.CompareFloat(-0.2, vectors[0].Value[0], 0.0001))
+	assert.True(t, common.CompareFloat(-0.3, vectors[1].Value[0], 0.0001))
+	assert.True(t, common.CompareFloat(-0.1, vectors[3].Value[0], 0.0001))
+}
+
+func TestAdamOptimizer(t *testing.T) {
+	d1 := []int64{2, 3}
+	v1 := []float32{1.0, 2.0, 3.0, 4.0, 5.0, 6.0}
+	t1 := common.Tensor{"t1", v1, d1, nil}
+
+	d2 := []int64{2, 2}
+	v2 := []float32{1.0, 2.0, 1.1, 2.2}
+	t2 := common.Tensor{"t2", v2, d2, nil}
+
+	p := NewParameter()
+	p.NonEmbeddingParam["t1"] = &t1
+	p.NonEmbeddingParam["t2"] = &t2
+
+	gv1 := []float32{1.0, 1.0, 1.0, 1.0, 1.0, 1.0}
+	gv2 := []float32{1.0, 1.0, 1.0, 1.0}
+	grad1 := common.Tensor{"t1", gv1, d1, nil}
+	grad2 := common.Tensor{"t2", gv2, d2, nil}
+	grads := []*common.Tensor{&grad1, &grad2}
+
+	opt := NewAdamOptimizer(0.1, 0.9, 0.999, 1e-8, false)
+	opt.step = 1
+
+	opt.InitNonEmbeddingParam("t1", d1)
+	opt.InitNonEmbeddingParam("t2", d2)
+
+	// test dense parameter update
+	err1 := opt.ApplyGradients(grads, p)
+	assert.Equal(t, opt.BaseOptimizer.lr, float32(0.1))
+	assert.Nil(t, err1)
+
+	ev1 := []float32{0.9255863187, 1.9255863187, 2.9255863187, 3.9255863187, 4.9255863187, 5.9255863187}
```

Comment:
It seems that we add these testing vectors directly. So are these vectors calculated by handcraft or by another tool like NumPy?

Reply:
by python